ehrql/query_engines/mssql.py

```diff
@@ -151,27 +153,86 @@ def create_inline_table(self, columns, rows):
+        https://learn.microsoft.com/en-us/sql/t-sql/queries/select-group-by-transact-sql?view=sql-server-ver16
+        """
+        all_sum_overs = [
+            sqlalchemy.func.sum(results_query.c[grouped_sum.denominator]).label("den")
+        ]
+        grouping_sets = []
+
+        for i, numerator in enumerate(grouped_sum.numerators):
+            all_sum_overs.append(
+                sqlalchemy.func.sum(results_query.c[numerator]).label(f"num_{i}"),
+            )
+            grouping_set = [
+                results_query.c[group_by_col]
+                for group_by_col in grouped_sum.group_bys[i]
+            ]
+            grouping_sets.append(grouping_set)
+        all_group_bys = ordered_set(iter_flatten(grouping_sets))
+        grouping_sets = [
+            sqlalchemy.tuple_(*grouping_set) for grouping_set in grouping_sets
+        ]
+        measures_query = sqlalchemy.select(
+            *all_sum_overs,
+            *all_group_bys,
+            sqlalchemy.func.grouping_id(*all_group_bys).label("grp_id"),
+        ).group_by(sqlalchemy.func.grouping_sets(*grouping_sets))
+
+        measures_results_table = temporary_table_from_query(
+            # Write the measures results to a global temporary table
+            f"##measures_results_{self.global_unique_id}",
+            measures_query,
+            index_col="num_0",
+        )
+        return [sqlalchemy.select(measures_results_table)]
+
     def get_queries(self, dataset):
         results_queries = super().get_queries(dataset)
         # Write results to temporary tables and select them from there. This allows us
         # to use more efficient/robust mechanisms to retrieve the results.
         select_queries = []
         for n, results_query in enumerate(results_queries, start=1):
+            index_col = "patient_id"
             results_table = temporary_table_from_query(
                 # The double `##` prefix here makes this a global temporary table, i.e.
                 # one accessible to other sessions, hence we need a globally unique
                 # name. This allows us to use a separate connection to retrieve results
                 # which gives us more robust retries.
                 f"##results_{self.global_unique_id}_{n}",
                 results_query,
-                index_col="patient_id",
+                index_col=index_col,
             )
             select_queries.append(sqlalchemy.select(results_table))
+
         return select_queries
 
     def get_results_stream(self, dataset):
         results_queries = self.get_queries(dataset)
 
+        if dataset.measures:
+            assert len(results_queries) == 1, (
+                "Measures queries can only be applied to a single results table"
+            )
+            results_query_table = results_queries[0].get_final_froms()[0]
+            results_queries = self.get_measure_queries(
+                dataset.measures, results_query_table
+            )
```

Comment:
I think you can probably get away without doing anything special here and just call `subquery()` on the results query as you do for the SQLite version. It does make the eventual query a bit convoluted because we're doing something like:
```sql
SELECT complicated_grouping_stuff() FROM
(
  SELECT * FROM some_table 
) AS subquery
```
But I'm pretty sure the query planner will see straight through that and will behave no differently than if we'd written:
```sql
SELECT complicated_grouping_stuff() FROM some_table
```

If it does turn out to make a difference then we can always add code to the BaseSQL engine which spots this specific pattern and turns it into the simpler one.